qa/workunits/rbd/rbd_mirror_helpers.sh

```diff
@@ -402,6 +406,82 @@ cleanup()
     exit ${error_code}
 }
 
+print_stacktrace() {
+    local frame=1 LINE SUB FILE
+    while read -r LINE SUB FILE < <(caller "$frame"); do
+      printf '[%s]  %s @ %s:%s\n' "$((frame++))" "${SUB}" "${FILE}" "${LINE}" 1>&2
+    done
+}
+
+run_cmd_internal() {
+    local exit_on_failure=$1 ; shift
+    local as_admin=$1 ; shift
+    local rc
+    local frame=0 LINE SUB FILE
+
+    if [ -n "${RBD_MIRROR_SHOW_CLI_CMD}" ]; then
+        if [ 'true' = "${as_admin}" ]; then
+            echo "CEPH_ARGS=''" "$@"
+        else
+            echo "CEPH_ARGS='--id ${CEPH_ID}'" "$@"
+        fi
+    fi
+
+    if [ 'true' = "${as_admin}" ]; then
+        export CEPH_ARGS=''
+    else
+        export CEPH_ARGS="--id ${CEPH_ID}"
+    fi
+
+    echo "$@" >> "${TEMPDIR}/rbd-mirror.cmd.log"
+
+    # Don't exit immediately if the command exits with a non-zero status.
+    set +e
+    eval $@ >"${CMD_STDOUT}" 2>"${CMD_STDERR}"
+    rc=$?
+    set -e
+
+    if [ -n "${RBD_MIRROR_SHOW_CLI_CMD}" ]; then
+        cat "$CMD_STDOUT"
+        cat "$CMD_STDERR" 1>&2
+    fi
+
+    if [ -n "${RBD_MIRROR_SAVE_CLI_OUTPUT}" ]; then
+        if [ 'true' = "${as_admin}" ]; then
+            echo "CEPH_ARGS=''" "$@" >> "${TEMPDIR}/${RBD_MIRROR_SAVE_CLI_OUTPUT}"
+        else
+            echo "CEPH_ARGS='--id ${CEPH_ID}'" "$@" >> "${TEMPDIR}/${RBD_MIRROR_SAVE_CLI_OUTPUT}"
+        fi
+        cat "$CMD_STDOUT" >> "${TEMPDIR}/${RBD_MIRROR_SAVE_CLI_OUTPUT}"
+        cat "$CMD_STDERR" >> "${TEMPDIR}/${RBD_MIRROR_SAVE_CLI_OUTPUT}"
+    fi
+
+    if [ 0 = $rc ] ; then
+        return 0
+    fi
+
+    if [ 'true' = "${exit_on_failure}" ]; then
+        echo -e "${RED}ERR: rc=" $rc 1>&2
+        print_stacktrace
+        echo -e "${NO_COLOUR}"
+        exit $rc
+    else
+        local frame=1 LINE SUB FILE
+
+        if [ -n "${RBD_MIRROR_SHOW_CLI_CMD}" ]; then
+            echo "ERR: rc=" $rc 1>&2
+            read -r LINE SUB FILE < <(caller "$frame")
+            printf "ERR: Non-fatal failure at: %s:%s %s()\n" "${FILE}" "${LINE}" "${SUB}" 1>&2
+        fi
+
+        return $rc
+    fi
+}
+
+run_cmd() {
```

Comment:
Can we avoid bringing in this framework, `run_cmd` `run_cmd_internal` `print_stack_trace` for `wip-rbd-cgsm-base`,  to keep this change simple?